User, Login JWT function

tranquoc2002 opened the pull request.

tranquoc2002 committed:
User, Login JWT function

tranquoc2002 committed:
add verify token

johntheagent2 commented on src/main/java/com/example/BookStore/controller/AccountController.java:
```diff
@@ -0,0 +1,60 @@
+    private AccountService accountService;
+
+    @PostMapping("/create")
+    public ApiResponse<Account> createAccount(@RequestBody @Valid CreateAccountRequest request){
```
không trả về nguyên cái Entity, chỉ trả DTO

johntheagent2 commented on src/main/java/com/example/BookStore/controller/AccountController.java:
```diff
@@ -0,0 +1,60 @@
+
+    @PostMapping("/create")
+    public ApiResponse<Account> createAccount(@RequestBody @Valid CreateAccountRequest request){
+        ApiResponse<Account> apiResponse = new ApiResponse<>();
```
không có tạo chay cái ApiResponse ở mỗi controller

johntheagent2 commented on src/main/java/com/example/BookStore/dto/request/AuthenticationRequest.java:
```diff
@@ -0,0 +1,18 @@
+@Builder
+public class AuthenticationRequest {
+    private String username;
+    private String password;
```
validate username, password. VD: kí tự đặc biệt, chữ hoa, chữ thường, số, max 8, min 6,..

johntheagent2 commented on src/main/java/com/example/BookStore/dto/response/ApiResponse.java:
```diff
@@ -0,0 +1,15 @@
+@Builder
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class ApiResponse <T> {
+    private int code = 1000;
```
1000 có nghĩa gì? sao không trả về HTTP status code?

johntheagent2 commented on src/main/java/com/example/BookStore/exception/ErrorCode.java:
```diff
@@ -0,0 +1,22 @@
+    USER_NOT_EXISTED(1002, "User not existed"),
+    USERNAME_INVALID(1003, "Username must be at least 3 characters"),
+    PASSWORD_INVALID(1004, "Password must be at least 8 characters"),
+    UNAUTHENTICATED(1005, "Unauthenticated")
```
mã lỗi không có nghĩa. Tại sao lại 999, 1001, 1002? Dùng HTTP Status code

johntheagent2 commented on src/main/java/com/example/BookStore/service/AccountService.java:
```diff
@@ -0,0 +1,67 @@
+@Service
+@Slf4j
+public class AccountService {
+    @Autowired
```
không dùng @Autowired. Dùng constructor injection

johntheagent2 commented on src/main/java/com/example/BookStore/service/AccountService.java:
```diff
@@ -0,0 +1,67 @@
+        account.setPassword(request.getPassword());
+        account.setFirstName(request.getFirstName());
+        account.setLastName(request.getLastName());
+        account.setDateOfBirth(request.getDateOfBirth());
```
nếu request cái field nó null thì sao đã validate chưa

johntheagent2 commented on src/main/java/com/example/BookStore/service/AuthenticationService.java:
```diff
@@ -0,0 +1,61 @@
+                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_EXISTED));
+
+        PasswordEncoder passwordEncoder = new BCryptPasswordEncoder(10);
+        boolean authenticated = passwordEncoder.matches(request.getPassword(), account.getPassword());
```
coi lại flow spring security. cái này Spring Security nó handle cho mình rồi

tranquoc2002 committed:
fixed 7/8 converstations

- Updated all AccountController not return to Entity (replace custom ApiResponse)
- Add more Validations in Request dtos
- Add DTO and Mapper for response
- Edit ErrorCode using HttpsStatus code

johntheagent2 commented on src/main/java/com/example/BookStore/controller/AccountController.java:
```diff
@@ -20,37 +23,55 @@ public class AccountController {
     @Autowired
```
không dùng autowired nữa. Ngoài ra không gọi trực tiếp cái service class. Best practice là tạo service implement cái service thí dụ như là AccountServiceImpl implements AccountService xong mới gọi private final AccountService trong Controller

tranquoc2002 committed:
Add Registration function, change some code block

Account -> Entity Base
Add User extend from Account
working Registration for User
change some code in Login function to work with User

johntheagent2 commented on src/main/java/com/example/BookStore/config/SecurityConfig.java:
```diff
@@ -0,0 +1,64 @@
+        httpSecurity.csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(request -> request
+                        .requestMatchers("/accounts/**").hasRole("USER")
+                        .requestMatchers("/auth/**").permitAll()
```
coi lại cách tổ chức đường dẫn nha. Nên nhớ là mình đang làm api nên là viết cái url nó dễ sử dụng, ít phải thêm mới thí dụ như là:

1. API public: /api/v1/public/
2. API user: /api/v1/user/
3. API admin: /api/1/admin/

johntheagent2 commented on src/main/java/com/example/BookStore/controller/RegistrationController.java:
```diff
@@ -0,0 +1,29 @@
+
+    @PostMapping
+    public ResponseEntity<String> register(@RequestBody @Valid RegistrationRequest request) {
+        return ResponseEntity.ok(userService.registerUser(request));
```
mấy này không gửi gì về thì cứ để  Void  là được không cần String đăng kí thành công gì gì đâu. Status 200 là biết thành công rồi

johntheagent2 commented on src/main/java/com/example/BookStore/config/SecurityConfig.java:
```diff
@@ -0,0 +1,64 @@
+                .authorizeHttpRequests(request -> request
+                        .requestMatchers("/accounts/**").hasRole("USER")
+                        .requestMatchers("/auth/**").permitAll()
+                        .requestMatchers("/registration/**").permitAll()
```
coi lại cách tổ chức đường dẫn nha. Nên nhớ là mình đang làm api nên là viết cái url nó dễ sử dụng, ít phải thêm mới thí dụ như là:

API public: /api/v1/public/
API user: /api/v1/user/
API admin: /api/1/admin/

johntheagent2 commented on src/main/java/com/example/BookStore/regtoken/RegToken.java:
```diff
@@ -0,0 +1,49 @@
+@Data
+@RequiredArgsConstructor
+@Entity
+public class RegToken {
```
thí dụ về sau làm những cái như reset password, change mail, forgot password thì sẽ tạo một cái bảng mới hả? Sao không để là OTP rồi có thêm một field để phân biệt tính năng như là field method rồi cho giá trị field như là CHANGE_PASWORD, RESET_PASSWORD,...

johntheagent2 commented on src/main/java/com/example/BookStore/repo/RegTokenRepository.java:
```diff
@@ -0,0 +1,24 @@
+public interface RegTokenRepository extends JpaRepository<RegToken, Long> {
+    Optional<RegToken> findByToken(String token);
+
+    @Transactional
```
không để transactional trong repo mà để vào service. Vì Spring JPA có hỗ trợ cái này ở repo rồi

johntheagent2 commented on src/main/java/com/example/BookStore/repo/UserRepository.java:
```diff
@@ -0,0 +1,19 @@
+public interface UserRepository extends JpaRepository<User, Long> {
+    Optional<User> findByEmail(String email);
+
+    @Transactional
```
tương tự transactional

johntheagent2 commented on src/main/java/com/example/BookStore/service/AuthenticationService.java:
```diff
@@ -40,17 +36,24 @@ public IntrospectTokenResponse introspect(IntrospectTokenResquest resquest){
+                }
+            }
+        }
+        if (!passwordEncoder.matches(request.getPassword(), userDetails.getPassword())) {
```
không dùng cách này để authen. Dùng AuthenticationManager. Nó có hàm authenticate. Không cần check tay như này

johntheagent2 commented on src/main/java/com/example/BookStore/controller/AccountController.java:
```diff
@@ -0,0 +1,79 @@
+
+    }
+    @PutMapping("/{id}/update")
+    public ResponseEntity<ApiResponse> updateAccount(@PathVariable Long id, @RequestBody @Valid UpdateAccountRequest request){
```
không được truyền id vào như thế này đối với user

johntheagent2 commented on src/main/java/com/example/BookStore/service/Impl/AccountServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+    }
+    @Override
+    public AccountDTO updateAccount(Long id, UpdateAccountRequest request){
+        Account existAccount = accountRepository.findById(id).orElseThrow(
```
t thấy account base entity của m nó có email là unique. Xong dùng SecurityContextHolder để lấy cái email/username gì đó ra để tìm

johntheagent2 commented on src/main/java/com/example/BookStore/service/Impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+    }
+    @Override
+    public String confirmUserToken(String token) {
+        RegToken regToken = regTokenService.getToken(token)
```
nếu người dùng khác sử dụng token của t thì người dùng khác sẽ được verify otp hay t được verify otp? Làm sao biết token nào của ai, ai đang verify?

johntheagent2 commented on src/main/java/com/example/BookStore/service/Impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+    private final BCryptPasswordEncoder passwordEncoder;
+
+    @Override
+    public String registerUser(RegistrationRequest request) {
```
sao không tách ra registration service riêng để nó handle mấy cái logic đăng kí rồi thằng UserService này để tạo lưu thôi. Tuân thủ SOLID nha, 1 method thực hiện chức năng của 1 mình nó

johntheagent2 commented on src/main/java/com/example/BookStore/service/Impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+        userRepository.save(user);
+
+
+        String token = UUID.randomUUID().toString();
```
thí dụ như này thì nên tách ra thành một cái class common nha thay vì lúc nào cần token thì phải ghi ra thì có thể tạo static ở một class ví dụ như Common. Rồi cần thì Common::generateUUIDToken

johntheagent2 commented on src/main/java/com/example/BookStore/service/Impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+            throw new AppException(ErrorCode.ACCOUNT_ALREADY_ENABLED);
+        }
+
+        String newToken = UUID.randomUUID().toString();
```
tương tự nè. Ghi lại nhiều => mốt đổi thì phải kiếm để đổi thay vì đổi một chỗ
